src/plot_api/subroutines.js

```diff
@@ -240,6 +240,11 @@ function lsInner(gd) {
     }
 
     function yLinePathFree(x) {
+        if(typeof(ya.shift) === 'number') {
+            x += ya.shift;
+        } else if(ya._shift !== undefined) {
+            x += ya._shift;
+        }
```

Comment:
@alexcjohnson @archmoj I'm also curious to hear your thoughts on the bug mentioned in the PR description. I've noticed that the y-axis lines with `shift=true` aren't offset correctly in some cases when the `xaxis` domain is specified. I think that this is because of automargin and the timing of when this function (which I've modified to shift the axis lines). This function relies on the `ax._shift` parameter, which is only set during axis drawing (`axes.DrawOne`). It works when each axis is drawn >1x, which is the case when `automargin` needs to trigger a redraw to resize the margins. If the domain is constrained such that there's already enough margin space on the first draw, then each axis is only drawn once and the `ax._shift` parameter won't have been updated by the time the final version of the axis line is drawn. If this understanding is correct, I'm not sure how to get around this. I'll need to keep setting `ax._shift` within `axes.DrawOne` as this is when the axis depths are calculated. I also don't totally follow the `automargin` redrawing logic -- is there are way to trigger that for this condition?

Reply:
Oh that's a really interesting bug... nice catch! Let's make sure there's a mock that hits this condition, so once we fix it we'll be sure it stays that way!

I guess we have an implicit assumption built into the way we draw axes that only the margins (and the axis attributes) can impact where the axis lines are drawn. And I think that was a valid assumption... until now 😅 

The *best* way to solve this would be to move axis line drawing out of `layoutStyles` and into `axes.DrawOne`. That would also make that particular part of the code much easier to follow, as everything about drawing a given axis would be in one place. It'd be a slight overhead in a pretty hot code path (we redraw the axis a LOT during certain zoom/pan interactions) but I don't think that's a big deal, and anyway we should be able to intelligently short-circuit it if we think it's important enough.

Does that seem tractable, or too ambitious a refactor?

Reply:
That looks about right. You should always be able to find the subplot object if you need it, even from the axis drawing routine. I'm not sure it matters too much where we store info / object references inside `_fullLayout`, we can do it whichever way is easiest. But it is definitely important not to change what container these lines get drawn in, as that will affect their order on screen, which is important mainly for inset plots.

Reply:
@alexcjohnson I have this in progress [here](https://github.com/plotly/plotly.js/pull/6408). The implementation I have so far works well in solving the bug but seems to have introduced some problems with drawing mirrored axes on linked subplots. My initial hunch is that this is due to the sequence of events when drawing -- I'm wondering if it's the case that not all necessary subplots have been created at the time of axis drawing. 

It might be the case that this refactor is in fact a bit ambitious for the scope of this PR. I've added [this change](https://github.com/plotly/plotly.js/pull/6334/commits/e56020db8ee2a9c9c1e374cc9ef5930ac426529e) which is a workaround to trigger a redraw of the plot when axes are shifted. Do you think this is reasonable for now?